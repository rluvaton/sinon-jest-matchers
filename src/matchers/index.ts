export { sinonToBeCalled } from './sinonToBeCalled';
export { sinonToBeCalledTimes } from './sinonToBeCalledTimes';
export { sinonToBeCalledWith } from './sinonToBeCalledWith';
export { sinonLastCalledWith } from './sinonLastCalledWith';
export { sinonLastReturnedWith } from './sinonLastReturnedWith';
export { sinonNthCalledWith } from './sinonNthCalledWith';
export { sinonToReturn } from './sinonToReturn';
export { sinonToReturnTimes } from './sinonToReturnTimes';
export { sinonToReturnWith } from './sinonToReturnWith';
