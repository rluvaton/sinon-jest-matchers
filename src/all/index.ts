import * as matchers from '../matchers';

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
const jestExpect = global.expect;

if (jestExpect !== undefined) {
  jestExpect.extend(matchers);
} else {
  throw new Error(
    "Unable to find Jest's global expect. " +
      'Please check you have added sinon-jest-matchers correctly to your jest configuration. ' +
      'See https://github.com/rluvaton/sinon-jest-matchers#setup for help.',
  );
}
