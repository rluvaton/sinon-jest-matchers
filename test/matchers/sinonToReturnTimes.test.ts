import sinon from 'sinon';
import * as matchers from '../../src';

expect.extend(matchers);

describe('sinonToReturnTimes', () => {
  test('only accepts a number argument', () => {
    const fn = sinon.spy(() => 42);
    fn();
    expect(fn).sinonToReturnTimes(1);

    [{}, [], true, 'a', new Map(), () => {}].forEach(value => {
      // @ts-expect-error we test here invalid arguments
      expect(() => expect(fn).sinonToReturnTimes(value)).toThrowErrorMatchingSnapshot();
    });
  });

  test('.not only accepts a number argument', () => {
    const fn = sinon.spy(() => 42);
    expect(fn).not.sinonToReturnTimes(2);

    [{}, [], true, 'a', new Map(), () => {}].forEach(value => {
      // @ts-expect-error we test here invalid arguments
      expect(() => expect(fn).not.sinonToReturnTimes(value)).toThrowErrorMatchingSnapshot();
    });
  });

  test('passes if function returned equal to expected times', () => {
    const fn = sinon.spy(() => 42);
    fn();
    fn();

    expect(fn).sinonToReturnTimes(2);

    expect(() => expect(fn).not.sinonToReturnTimes(2)).toThrowErrorMatchingSnapshot();
  });

  test('calls that return undefined are counted as returns', () => {
    const fn = sinon.spy(() => undefined);
    fn();
    fn();

    expect(fn).sinonToReturnTimes(2);

    expect(() => expect(fn).not.sinonToReturnTimes(2)).toThrowErrorMatchingSnapshot();
  });

  test('.not passes if function returned more than expected times', () => {
    const fn = sinon.spy(() => 42);
    fn();
    fn();
    fn();

    expect(fn).sinonToReturnTimes(3);
    expect(fn).not.sinonToReturnTimes(2);

    expect(() => expect(fn).sinonToReturnTimes(2)).toThrowErrorMatchingSnapshot();
  });

  test('.not passes if function called less than expected times', () => {
    const fn = sinon.spy(() => 42);
    fn();

    expect(fn).sinonToReturnTimes(1);
    expect(fn).not.sinonToReturnTimes(2);

    expect(() => expect(fn).sinonToReturnTimes(2)).toThrowErrorMatchingSnapshot();
  });

  test('calls that throw are not counted', () => {
    const fn = sinon.spy(causeError => {
      if (causeError) {
        throw new Error('Error!');
      }

      return 42;
    });

    fn(false);

    try {
      fn(true);
    } catch {
      // ignore error
    }

    fn(false);

    expect(fn).not.sinonToReturnTimes(3);

    expect(() => expect(fn).sinonToReturnTimes(3)).toThrowErrorMatchingSnapshot();
  });

  // Bug in sinon - https://github.com/sinonjs/sinon/issues/2471
  test.skip('calls that throw undefined are not counted', () => {
    const fn = sinon.spy(causeError => {
      if (causeError) {
        // eslint-disable-next-line no-throw-literal
        throw undefined;
      }

      return 42;
    });

    fn(false);

    try {
      fn(true);
    } catch {
      // ignore error
    }

    fn(false);

    expect(fn).sinonToReturnTimes(2);

    expect(() => expect(fn).not.sinonToReturnTimes(2)).toThrowErrorMatchingSnapshot();
  });

  test('includes the custom mock name in the error message', () => {
    const fn = sinon.spy(() => 42);
    fn.named('named-mock');
    fn();
    fn();

    expect(fn).sinonToReturnTimes(2);

    expect(() => expect(fn).sinonToReturnTimes(1)).toThrowErrorMatchingSnapshot();
  });
});
