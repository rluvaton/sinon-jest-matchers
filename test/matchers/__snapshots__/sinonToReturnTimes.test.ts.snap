// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sinonToReturnTimes .not only accepts a number argument 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

[1mMatcher error[22m: [32mexpected[39m value must be a non-negative integer

Expected has type:  object
Expected has value: [32m{}[39m"
`;

exports[`sinonToReturnTimes .not only accepts a number argument 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

[1mMatcher error[22m: [32mexpected[39m value must be a non-negative integer

Expected has type:  array
Expected has value: [32m[][39m"
`;

exports[`sinonToReturnTimes .not only accepts a number argument 3`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

[1mMatcher error[22m: [32mexpected[39m value must be a non-negative integer

Expected has type:  boolean
Expected has value: [32mtrue[39m"
`;

exports[`sinonToReturnTimes .not only accepts a number argument 4`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

[1mMatcher error[22m: [32mexpected[39m value must be a non-negative integer

Expected has type:  string
Expected has value: [32m\\"a\\"[39m"
`;

exports[`sinonToReturnTimes .not only accepts a number argument 5`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

[1mMatcher error[22m: [32mexpected[39m value must be a non-negative integer

Expected has type:  map
Expected has value: [32mMap {}[39m"
`;

exports[`sinonToReturnTimes .not only accepts a number argument 6`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

[1mMatcher error[22m: [32mexpected[39m value must be a non-negative integer

Expected has type:  function
Expected has value: [32m[Function anonymous][39m"
`;

exports[`sinonToReturnTimes .not passes if function called less than expected times 1`] = `
"[2mexpect.[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

Expected number of returns: [32m2[39m
Received number of returns: [31m1[39m"
`;

exports[`sinonToReturnTimes .not passes if function returned more than expected times 1`] = `
"[2mexpect.[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

Expected number of returns: [32m2[39m
Received number of returns: [31m3[39m"
`;

exports[`sinonToReturnTimes calls that return undefined are counted as returns 1`] = `
"[2mexpect.[22mnot[2m.[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

Expected number of returns: not [32m2[39m"
`;

exports[`sinonToReturnTimes calls that throw are not counted 1`] = `
"[2mexpect.[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

Expected number of returns: [32m3[39m
Received number of returns: [31m2[39m
Received number of calls:   [31m3[39m"
`;

exports[`sinonToReturnTimes includes the custom mock name in the error message 1`] = `
"[2mexpect([22m[31mnamed-mock[39m[2m).[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

Expected number of returns: [32m1[39m
Received number of returns: [31m2[39m"
`;

exports[`sinonToReturnTimes only accepts a number argument 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

[1mMatcher error[22m: [32mexpected[39m value must be a non-negative integer

Expected has type:  object
Expected has value: [32m{}[39m"
`;

exports[`sinonToReturnTimes only accepts a number argument 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

[1mMatcher error[22m: [32mexpected[39m value must be a non-negative integer

Expected has type:  array
Expected has value: [32m[][39m"
`;

exports[`sinonToReturnTimes only accepts a number argument 3`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

[1mMatcher error[22m: [32mexpected[39m value must be a non-negative integer

Expected has type:  boolean
Expected has value: [32mtrue[39m"
`;

exports[`sinonToReturnTimes only accepts a number argument 4`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

[1mMatcher error[22m: [32mexpected[39m value must be a non-negative integer

Expected has type:  string
Expected has value: [32m\\"a\\"[39m"
`;

exports[`sinonToReturnTimes only accepts a number argument 5`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

[1mMatcher error[22m: [32mexpected[39m value must be a non-negative integer

Expected has type:  map
Expected has value: [32mMap {}[39m"
`;

exports[`sinonToReturnTimes only accepts a number argument 6`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

[1mMatcher error[22m: [32mexpected[39m value must be a non-negative integer

Expected has type:  function
Expected has value: [32m[Function anonymous][39m"
`;

exports[`sinonToReturnTimes passes if function returned equal to expected times 1`] = `
"[2mexpect.[22mnot[2m.[22msinonToReturnTimes[2m([22m[32mexpected[39m[2m)[22m

Expected number of returns: not [32m2[39m"
`;
